Userland/Libraries/LibCrypto/Checksum/CRC32.h

```diff
@@ -26,11 +32,183 @@ class CRC32 : public ChecksumFunction<u32> {
         update(data);
     }
 
-    virtual void update(ReadonlyBytes data) override;
-    virtual u32 digest() override;
+    void update(ReadonlyBytes data) override;
+
+    u32 digest() override;
 
 private:
+#if !defined(__aarch64__) && !defined(__ARM_FEATURE_CRC32)
+#    if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+    static constexpr Array<Array<u32, 256>, 8> generate_table();
+    static constexpr auto m_table = generate_table();
+    static constexpr u32 single_byte_crc(u32 crc, u8 byte);
+#    else
+    static constexpr Array<u32, 256> generate_table();
+    static constexpr auto m_table = generate_table();
+#    endif
+#endif
+
     u32 m_state { ~0u };
 };
 
+#if defined(__aarch64__) && defined(__ARM_FEATURE_CRC32)
+
+template<u32 polynomial>
+void CRC32<polynomial>::update(ReadonlyBytes span)
+{
+    // FIXME: Does this require runtime checking on rpi?
+    //        (Maybe the instruction is present on the rpi4 but not on the rpi3?)
+
+    u8 const* data = span.data();
+    size_t size = span.size();
+
+    while (size > 0 && (reinterpret_cast<FlatPtr>(data) & 7) != 0) {
+        m_state = __builtin_arm_crc32b(m_state, *data);
+        ++data;
+        --size;
+    }
+
+    auto* data64 = reinterpret_cast<u64 const*>(data);
+    while (size >= 8) {
+        m_state = __builtin_arm_crc32d(m_state, *data64);
+        ++data64;
+        size -= 8;
+    }
+
+    data = reinterpret_cast<u8 const*>(data64);
+    while (size > 0) {
+        m_state = __builtin_arm_crc32b(m_state, *data);
+        ++data;
+        --size;
+    }
+}
```

Comment:
This should only be used for the `0x04C11DB7` (ethernet, mpeg2, png, gzip, SATA) polynomial
https://developer.arm.com/documentation/dui0801/h/A64-General-Instructions/CRC32B--CRC32H--CRC32W--CRC32X
There are also instructions for the `0x1EDC6F41` (iSCSI,  ext4, btrfs) polynomial
https://developer.arm.com/documentation/dui0801/h/A64-General-Instructions/CRC32CB--CRC32CH--CRC32CW--CRC32CX
Which is also the one SSE4.2 provides

For that you can use partial specialization of the template